snap/gadget.go

```diff
@@ -234,6 +256,221 @@ func ReadGadgetInfo(info *Info, classic bool) (*GadgetInfo, error) {
 	return &gi, nil
 }
 
+func fmtIndexAndName(idx int, name string) string {
+	if name != "" {
+		return fmt.Sprintf("#%v (%q)", idx, name)
+	}
+	return fmt.Sprintf("#%v", idx)
+}
+
+func validateVolume(name string, vol *GadgetVolume) error {
+	if !validVolumeName.MatchString(name) {
+		return errors.New("invalid volume name")
+	}
+	if vol.Schema != "" && vol.Schema != "gpt" && vol.Schema != "mbr" {
+		return fmt.Errorf("invalid volume schema %q", vol.Schema)
+	}
+
+	structureNames := make(map[string]bool, len(vol.Structure))
+	for idx, s := range vol.Structure {
+		if err := validateVolumeStructure(&s, vol); err != nil {
+			return fmt.Errorf("invalid structure %s: %v", fmtIndexAndName(idx, s.Name), err)
+		}
+		if s.Name != "" {
+			structureNames[s.Name] = true
```

Comment:
can structures repeat the same name?

Reply:
Another piece that AFAIK is not handled by ubuntu-image. Nonetheless, we should check for it.

Reply:
Opened a PR to ubuntu-image with the fix too.